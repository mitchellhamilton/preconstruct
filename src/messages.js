// @flow

export let errors = {
  noEntryPoint:
    "no entrypoint was provided, please create a file at src/index.js",
  deniedWriteMainField:
    "changing the main field is required for preconstruct to build",
  invalidModuleField: "module field is invalid",
  invalidMainField: "main field is invalid",
  invalidUmdMainField: "umd:main field is invalid",
  invalidBrowserField: "browser field is invalid",
  invalidReactNativeField: "react-native field is invalid",
  umdNameNotSpecified:
    "the umd:main field is specified but a umdName option is not specified. please add it to the preconstruct field in your package.json",
  deniedWriteBrowserField:
    "building browser bundles for modules that include typeof window or typeof document is currently required"
};

import { createPromptConfirmLoader } from "./prompt";

export let confirms = {
  writeMainField: createPromptConfirmLoader(
    "preconstruct is going to change the main field in your package.json, are you okay with that?"
  ),
  writeModuleField: createPromptConfirmLoader(
    "would you like to generate module builds? this will write to the module field in your package.json"
  ),
  fixModuleField: createPromptConfirmLoader(
    "would you like to fix the module field?"
  ),
  writeUmdBuilds: createPromptConfirmLoader(
    "would you like to generate UMD builds?"
  ),
  shouldInstallObjectAssign: createPromptConfirmLoader(
    "Object.assign is polyfilled with object-assign to reduce bundle size when used with react. would you like to install object-assign automatically?"
  ),
  addBrowserField: createPromptConfirmLoader(
    "typeof window or typeof document is used in this package. would you like build seperate browser builds for better browser bundle sizes?"
  )
};

export let inputs = {
  getUmdName: "what should the name used for UMD bundles be?"
};

export let infos = {
  validMainField: "main field is valid",
  validModuleField: "module field is valid",
  validUmdMainField: "umd:main field is valid",
  validEntrypoint: "a valid entry point exists.",
  validBrowserField: "browser field is valid",
  validReactNativeField: "react-native field is valid"
};

export let successes = {
  validPackage: "package is valid!",
  startedWatching: "started watching!"
};
