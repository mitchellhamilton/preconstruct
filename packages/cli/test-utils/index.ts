import path from "path";
import * as fs from "fs-extra";
import globby from "globby";
import fixturez from "fixturez";
import spawn from "spawndamnit";
// @ts-ignore
import initHasher from "xxhash-wasm";
// @ts-ignore
import profiler from "v8-profiler-next";

let f = fixturez(__dirname);

require("chalk").enabled = false;
console.error = jest.fn();
console.log = jest.fn();

export let logMock = {
  log: (console.log as any) as jest.MockInstance<void, any>,
  error: (console.error as any) as jest.MockInstance<void, any>
};

afterEach(() => {
  logMock.log.mockReset();
  logMock.error.mockReset();
});

import init from "../src/init";
import { confirms } from "../src/messages";

let mockedConfirms = confirms as jest.Mocked<typeof confirms>;

export async function initBasic(directory: string) {
  mockedConfirms.writeMainField.mockReturnValue(Promise.resolve(true));
  mockedConfirms.writeModuleField.mockReturnValue(Promise.resolve(true));

  await init(directory);
  mockedConfirms.writeMainField.mockReset();
  mockedConfirms.writeModuleField.mockReset();
}

export function profile(name: string) {
  profiler.startProfiling(name, true);

  return () => {
    let profile = profiler.stopProfiling();

    new Promise<void>(resolve =>
      profile
        .export()
        .pipe(fs.createWriteStream(name + ".cpuprofile"))
        .on("finish", function() {
          profile.delete();
          resolve();
        })
    );
  };
}

function getPkgPath(tmpPath: string) {
  return path.join(tmpPath, "package.json");
}

export async function getPkg(
  filepath: string
): Promise<{ [key: string]: any }> {
  return JSON.parse(
    await fs.readFile(path.join(filepath, "package.json"), "utf-8")
  );
}

export async function modifyPkg(
  tmpPath: string,
  cb: (pkgJson: { [key: string]: any }) => unknown
) {
  let json = await getPkg(tmpPath);
  await cb(json);

  let pkgPath = getPkgPath(tmpPath);
  await fs.writeFile(pkgPath, JSON.stringify(json, null, 2));
}

export let createPackageCheckTestCreator = (
  doResult: (tmpPath: string) => Promise<void>
) => {
  let createTestCreator = (
    testFn: (testName: string, func: () => Promise<void>) => void
  ) => async (
    testName: string,
    entrypoints: { [key: string]: any },
    cb: (
      doThing: () => Promise<{ [key: string]: any }>,
      stuff: { dir: string }
    ) => Promise<void>
  ) => {
    testFn(testName, async () => {
      let tmpPath = f.copy("template-simple-package");
      let things = Object.keys(entrypoints);
      await Promise.all(
        things.map(async entrypointPath => {
          let content = entrypoints[entrypointPath];
          let filepath = path.join(tmpPath, entrypointPath, "package.json");
          await fs.ensureFile(filepath);
          await fs.writeFile(filepath, JSON.stringify(content, null, 2));
        })
      );

      await cb(
        async () => {
          await doResult(tmpPath);

          let newThings: Record<string, any> = {};

          await Promise.all(
            things.map(async entrypointPath => {
              newThings[entrypointPath] = JSON.parse(
                await fs.readFile(
                  path.join(tmpPath, entrypointPath, "package.json"),
                  "utf8"
                )
              );
            })
          );
          return newThings;
        },
        { dir: tmpPath }
      );
    });
  };
  let testFn = createTestCreator(test);
  // @ts-ignore
  testFn.only = createTestCreator(test.only);
  // @ts-ignore
  testFn.skip = createTestCreator(test.skip);
  return testFn as typeof testFn & { only: typeof testFn; skip: typeof testFn };
};

export async function snapshotDistFiles(tmpPath: string) {
  let distPath = path.join(tmpPath, "dist");
  let distFiles;
  try {
    distFiles = await fs.readdir(distPath);
  } catch (err) {
    if (err.code === "ENOENT") {
      throw new Error(distPath + " does not exist");
    }
    throw err;
  }

  await Promise.all(
    distFiles.map(async x => {
      expect(
        await fs.readFile(path.join(distPath, x), "utf-8")
      ).toMatchSnapshot(x);
    })
  );
}

export let stripHashes = async (chunkName: string) => {
  let { h64 } = await initHasher();

  let transformer = (pathname: string, content: string) => {
    return pathname.replace(new RegExp(`${chunkName}-[^\\.]+`, "g"), () => {
      return `chunk-this-is-not-the-real-hash-${h64(content)}`;
    });
  };
  return {
    transformPath: transformer,
    transformContent: (content: string) => {
      return content.replace(new RegExp(`${chunkName}-[^\\.]+`, "g"), () => {
        return "chunk-some-hash";
      });
    }
  };
};

export async function snapshotDirectory(
  tmpPath: string,
  {
    files = "js",
    filterPath = x => true,
    transformPath = x => x,
    transformContent = x => x
  }: {
    files?: "all" | "js";
    filterPath?: (path: string) => boolean;
    transformPath?: (path: string, contents: string) => string;
    transformContent?: (content: string) => string;
  } = {}
) {
  let paths = await globby(
    [`**/${files === "js" ? "*.js" : "*"}`, "!node_modules/**", "!yarn.lock"],
    {
      cwd: tmpPath
    }
  );

  await Promise.all(
    paths.filter(fp => filterPath(fp)).map(async x => {
      let content = transformContent(
        await fs.readFile(path.join(tmpPath, x), "utf-8")
      );
      if (x.endsWith(".json") && !x.endsWith("tsconfig.json")) {
        content = JSON.parse(content);
      }
      expect(content).toMatchSnapshot(transformPath(x, content));
    })
  );
}

export async function install(tmpPath: string) {
  await spawn("yarn", ["install"], { cwd: tmpPath });
}
