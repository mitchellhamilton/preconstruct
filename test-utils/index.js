// @flow
import path from "path";
import * as fs from "fs-extra";
import globby from "globby";
import fixturez from "fixturez";

let f = fixturez(__dirname);

require("chalk").enabled = false;
// $FlowFixMe
console.error = jest.fn();
// $FlowFixMe
console.log = jest.fn();

export let logMock = {
  log: ((console.log: any): JestMockFn<any, void>),
  error: ((console.error: any): JestMockFn<any, void>)
};

afterEach(() => {
  logMock.log.mockReset();
  logMock.error.mockReset();
});

import init from "../src/init";
import { confirms } from "../src/messages";

export async function initBasic(directory: string) {
  confirms.writeMainField.mockReturnValue(true);
  confirms.writeModuleField.mockReturnValue(true);
  confirms.writeUmdBuilds.mockReturnValue(false);

  await init(directory);
  confirms.writeMainField.mockReset();
  confirms.writeModuleField.mockReset();
  confirms.writeUmdBuilds.mockReset();
}

function getPkgPath(tmpPath: string) {
  return path.join(tmpPath, "package.json");
}

export async function getPkg(filepath: string): Object {
  return JSON.parse(
    await fs.readFile(path.join(filepath, "package.json"), "utf-8")
  );
}

export async function modifyPkg(tmpPath: string, cb: Object => mixed) {
  let json = await getPkg(tmpPath);
  await cb(json);

  let pkgPath = getPkgPath(tmpPath);
  await fs.writeFile(pkgPath, JSON.stringify(json, null, 2));
}

export let createPackageCheckTestCreator = (
  doResult: string => Promise<void>
) => {
  let createTestCreator = testFn => async (
    testName: string,
    entrypoints: { [key: string]: Object },
    cb: (doThing: () => Promise<{ [key: string]: Object }>) => Promise<void>
  ) => {
    testFn(testName, async () => {
      let tmpPath = f.copy("template-simple-package");
      let things = Object.keys(entrypoints);
      await Promise.all(
        things.map(async entrypointPath => {
          let content = entrypoints[entrypointPath];
          let filepath = path.join(tmpPath, entrypointPath, "package.json");
          await fs.ensureFile(filepath);
          await fs.writeFile(filepath, JSON.stringify(content, null, 2));
        })
      );

      await cb(async () => {
        await doResult(tmpPath);

        let newThings = {};

        await Promise.all(
          things.map(async entrypointPath => {
            newThings[entrypointPath] = JSON.parse(
              await fs.readFile(
                path.join(tmpPath, entrypointPath, "package.json"),
                "utf8"
              )
            );
          })
        );
        return newThings;
      });
    });
  };
  let testFn = createTestCreator(test);
  testFn.only = createTestCreator(test.only);
  testFn.skip = createTestCreator(test.skip);
  return testFn;
};

export async function snapshotDistFiles(tmpPath: string) {
  let distPath = path.join(tmpPath, "dist");
  let distFiles;
  try {
    distFiles = await fs.readdir(distPath);
  } catch (err) {
    if (err.code === "ENOENT") {
      throw new Error(distPath + " does not exist");
    }
    throw err;
  }

  await Promise.all(
    distFiles.map(async x => {
      expect(
        await fs.readFile(path.join(distPath, x), "utf-8")
      ).toMatchSnapshot(x);
    })
  );
}

export async function snapshotDirectory(
  tmpPath: string,
  files: "all" | "js" = "js"
) {
  let paths = await globby([`**/${files === "js" ? "*.js" : "*"}`], {
    cwd: tmpPath
  });

  await Promise.all(
    paths.map(async x => {
      expect(await fs.readFile(path.join(tmpPath, x), "utf-8")).toMatchSnapshot(
        x
      );
    })
  );
}
