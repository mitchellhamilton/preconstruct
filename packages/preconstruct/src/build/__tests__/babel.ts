import build from "../";
import fixturez from "fixturez";
import { confirms as _confirms } from "../../messages";
import { install, snapshotDirectory } from "../../../test-utils";

const f = fixturez(__dirname);

jest.setTimeout(20000);

jest.mock("../../prompt");

beforeEach(() => {
  jest.resetAllMocks();
});

let confirms = _confirms as jest.Mocked<typeof _confirms>;

// this test is causing to much frustration
test.skip("needs @babel/runtime allow install", async () => {
  let tmpPath = f.copy("use-babel-runtime");
  await install(tmpPath);
  confirms.shouldInstallBabelRuntime.mockReturnValue(Promise.resolve(true));

  await build(tmpPath);

  // TODO: investigate why this is called more than one time
  expect(confirms.shouldInstallBabelRuntime).toBeCalled();

  await snapshotDirectory(tmpPath, { files: "all" });
});
