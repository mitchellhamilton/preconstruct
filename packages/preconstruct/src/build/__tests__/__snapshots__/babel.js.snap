// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`needs @babel/runtime allow install: dist/use-babel-runtime.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));

var Bar = function Bar() {
  _classCallCheck(this, Bar);
};

var Foo =
/*#__PURE__*/
function (_Bar) {
  _inherits(Foo, _Bar);

  function Foo() {
    _classCallCheck(this, Foo);

    return _possibleConstructorReturn(this, _getPrototypeOf(Foo).apply(this, arguments));
  }

  return Foo;
}(Bar);

exports.Foo = Foo;
"
`;

exports[`needs @babel/runtime allow install: dist/use-babel-runtime.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./use-babel-runtime.cjs.prod.js\\");
} else {
  module.exports = require(\\"./use-babel-runtime.cjs.dev.js\\");
}
"
`;

exports[`needs @babel/runtime allow install: dist/use-babel-runtime.cjs.prod.js 1`] = `
"\\"use strict\\";

function _interopDefault(ex) {
  return ex && \\"object\\" == typeof ex && \\"default\\" in ex ? ex.default : ex;
}

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

var _possibleConstructorReturn = _interopDefault(require(\\"@babel/runtime/helpers/possibleConstructorReturn\\")), _getPrototypeOf = _interopDefault(require(\\"@babel/runtime/helpers/getPrototypeOf\\")), _inherits = _interopDefault(require(\\"@babel/runtime/helpers/inherits\\")), _classCallCheck = _interopDefault(require(\\"@babel/runtime/helpers/classCallCheck\\")), Bar = function Bar() {
  _classCallCheck(this, Bar);
}, Foo = function(_Bar) {
  function Foo() {
    return _classCallCheck(this, Foo), _possibleConstructorReturn(this, _getPrototypeOf(Foo).apply(this, arguments));
  }
  return _inherits(Foo, Bar), Foo;
}();

exports.Foo = Foo;
"
`;

exports[`needs @babel/runtime allow install: dist/use-babel-runtime.esm.js 1`] = `
"import _possibleConstructorReturn from '@babel/runtime/helpers/esm/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/esm/getPrototypeOf';
import _inherits from '@babel/runtime/helpers/esm/inherits';
import _classCallCheck from '@babel/runtime/helpers/esm/classCallCheck';

var Bar = function Bar() {
  _classCallCheck(this, Bar);
};

var Foo =
/*#__PURE__*/
function (_Bar) {
  _inherits(Foo, _Bar);

  function Foo() {
    _classCallCheck(this, Foo);

    return _possibleConstructorReturn(this, _getPrototypeOf(Foo).apply(this, arguments));
  }

  return Foo;
}(Bar);

export { Foo };
"
`;

exports[`needs @babel/runtime allow install: package.json 1`] = `
"{
  \\"name\\": \\"use-babel-runtime\\",
  \\"version\\": \\"1.0.0\\",
  \\"main\\": \\"dist/use-babel-runtime.cjs.js\\",
  \\"module\\": \\"dist/use-babel-runtime.esm.js\\",
  \\"license\\": \\"MIT\\",
  \\"private\\": true,
  \\"dependencies\\": {
    \\"@babel/core\\": \\"^7.4.3\\",
    \\"@babel/preset-env\\": \\"^7.4.3\\",
    \\"@babel/runtime\\": \\"^7.4.5\\"
  }
}
"
`;

exports[`needs @babel/runtime allow install: src/index.js 1`] = `
"class Bar {}

export class Foo extends Bar {}
"
`;
